// levelup/lib/levelup.js
/* Copyright (c) 2012-2013 LevelUP contributors
 * See list at <https://github.com/rvagg/node-levelup#contributing>
 * MIT +no-false-attribs License
 * <https://github.com/rvagg/node-levelup/blob/master/LICENSE>
 */

var EventEmitter   = require('events').EventEmitter
  , inherits       = require('util').inherits
  , extend         = require('xtend')
  , prr            = require('prr')

  , errors         = require('./errors')
  , readStream     = require('./read-stream')
  , writeStream    = require('./write-stream')
  , util           = require('./util')

  , getOptions     = util.getOptions
  , defaultOptions = util.defaultOptions
  , getLevelDOWN   = util.getLevelDOWN


  , createLevelUP = function (location, options, callback) {

      // Possible status values:
      //  - 'new'     - newly created, not opened or closed
      //  - 'opening' - waiting for the database to be opened, post open()
      //  - 'open'    - successfully opened the database, available for use
      //  - 'closing' - waiting for the database to be closed, post close()
      //  - 'closed'  - database has been successfully closed, should not be
      //                 used except for another open() operation

      var status = 'new'
        , error
        , levelup

        , isOpen        = function () { return status == 'open' }
        , isOpening     = function () { return status == 'opening' }

        , dispatchError = function (error, callback) {
            return typeof callback == 'function'
              ? callback(error)
              : levelup.emit('error', error)
          }

        , getCallback = function (options, callback) {
            return typeof options == 'function' ? options : callback
          }

        , deferred = [ 'get', 'put', 'batch', 'del', 'approximateSize' ]
            .reduce(function (o, method) {
              o[method] = function () {
                var args = Array.prototype.slice.call(arguments)
                levelup.once('ready', function () {
                  levelup.db[method].apply(levelup.db, args)
                })
              }
              return o
            }, {})

      if (typeof options == 'function') {
        callback = options
        options  = {}
      }

      options = getOptions(levelup, options)

      if (typeof location != 'string') {
        error = new errors.InitializationError(
            'Must provide a location for the database')
        if (callback)
          return callback(error)
        throw error
      }

      function LevelUP (location, options) {
        EventEmitter.call(this)
        this.setMaxListeners(Infinity)

        this.options = extend(defaultOptions, options)
        // set this.location as enumerable but not configurable or writable
        prr(this, 'location', location, 'e')
      }

      inherits(LevelUP, EventEmitter)

      LevelUP.prototype.open = function (callback) {
        var self = this
          , dbFactory
          , db

        if (isOpen()) {
          if (callback) {
            process.nextTick(function () { callback(null, self) })
          }
          return self
        }

        if (isOpening()) {
          return callback && levelup.once(
              'open'
            , function () { callback(null, self) }
          )
        }

        levelup.emit('opening')

        status = 'opening'
        self.db = deferred

        dbFactory = levelup.options.db || getLevelDOWN()
        db        = dbFactory(levelup.location)

        db.open(levelup.options, function (err) {
          if (err) {
            err = new errors.OpenError(err)
            return dispatchError(err, callback)
          } else {
            levelup.db = db
            status = 'open'
            if (callback)
              callback(null, levelup)
            levelup.emit('open')
            levelup.emit('ready')
          }
        })
      }

      LevelUP.prototype.close = function (callback) {
        if (isOpen()) {
          status = 'closing'
          this.db.close(function () {
            status = 'closed'
            levelup.emit('closed')
            if (callback)
              callback.apply(null, arguments)
          })
          levelup.emit('closing')
          this.db = null
        } else if (status == 'closed' && callback) {
          callback()
        } else if (status == 'closing' && callback) {
          levelup.once('closed', callback)
        } else if (isOpening()) {
          levelup.once('open', function () {
            levelup.close(callback)
          })
        }
      }

      LevelUP.prototype.isOpen = function () { return isOpen() }

      LevelUP.prototype.isClosed = function () { return (/^clos/).test(status) }

      LevelUP.prototype.get = function (key_, options, callback) {
        var key
          , err

        callback = getCallback(options, callback)

        if (typeof callback != 'function') {
          err = new errors.ReadError('get() requires key and callback arguments')
          return dispatchError(err)
        }

        if (!isOpening() && !isOpen()) {
          err = new errors.ReadError('Database is not open')
          return dispatchError(err, callback)
        }

        options = util.getOptions(levelup, options)
        key = util.encodeKey(key_, options)

        options.asBuffer = util.isValueAsBuffer(options)

        this.db.get(key, options, function (err, value) {
          if (err) {
            if ((/notfound/i).test(err)) {
              err = new errors.NotFoundError(
                  'Key not found in database [' + key_ + ']', err)
            } else {
              err = new errors.ReadError(err)
            }
            return dispatchError(err, callback)
          }
          if (callback) {
            try {
              value = util.decodeValue(value, options)
            } catch (e) {
              return callback(new errors.EncodingError(e))
            }
            callback(null, value)
          }
        })
      }

      LevelUP.prototype.put = function (key_, value_, options, callback) {
        var err
          , key
          , value

        callback = getCallback(options, callback)

        if (key_ === null || key_ === undefined
              || value_ === null || value_ === undefined) {
          err = new errors.WriteError('put() requires key and value arguments')
          return dispatchError(err, callback)
        }

        if (!isOpening() && !isOpen()) {
          err = new errors.WriteError('Database is not open')
          return dispatchError(err, callback)
        }

        options = getOptions(levelup, options)
        key     = util.encodeKey(key_, options)
        value   = util.encodeValue(value_, options)

        this.db.put(key, value, options, function (err) {
          if (err) {
            err = new errors.WriteError(err)
            return dispatchError(err, callback)
          } else {
            levelup.emit('put', key_, value_)
            if (callback)
              callback()
          }
        })
      }

      LevelUP.prototype.del = function (key_, options, callback) {
        var err
          , key

        callback = getCallback(options, callback)

        if (key_ === null || key_ === undefined) {
          err = new errors.WriteError('del() requires a key argument')
          return dispatchError(err, callback)
        }

        if (!isOpening() && !isOpen()) {
          err = new errors.WriteError('Database is not open')
          return dispatchError(err, callback)
        }

        options = getOptions(levelup, options)
        key     = util.encodeKey(key_, options)

        this.db.del(key, options, function (err) {
          if (err) {
            err = new errors.WriteError(err)
            return dispatchError(err, callback)
          } else {
            levelup.emit('del', key_)
            if (callback)
              callback()
          }
        })
      }

      function Batch (db) {
        this.batch = db.batch()
        this.ops = []
      }

      Batch.prototype.put = function (key_, value_, options) {
        options = getOptions(levelup, options)

        var key   = util.encodeKey(key_, options)
          , value = util.encodeValue(value_, options)

        try {
          this.batch.put(key, value)
        } catch (e) {
          throw new errors.WriteError(e)
        }
        this.ops.push({ type : 'put', key : key, value : value })

        return this
      }

      Batch.prototype.del = function (key_, options) {
        options = getOptions(levelup, options)
        var key     = util.encodeKey(key_, options)

        try {
          this.batch.del(key)
        } catch (err) {
          throw new errors.WriteError(err)
        }
        this.ops.push({ type : 'del', key : key })

        return this
      }

      Batch.prototype.clear = function () {
        try {
          this.batch.clear()
        } catch (err) {
          throw new errors.WriteError(err)
        }

        this.ops = []
        return this
      }

      Batch.prototype.write = function (callback) {
        var ops = this.ops
        try {
          this.batch.write(function (err) {
            if (err)
              return dispatchError(new errors.WriteError(err), callback)
            levelup.emit('batch', ops)
            if (callback)
              callback()
          })
        } catch (err) {
          throw new errors.WriteError(err)
        }
      }

      LevelUP.prototype.batch = function (arr_, options, callback) {
        var keyEnc
          , valueEnc
          , err
          , arr

        if (!arguments.length)
          return new Batch(this.db)

        callback = getCallback(options, callback)

        if (!Array.isArray(arr_)) {
          err = new errors.WriteError('batch() requires an array argument')
          return dispatchError(err, callback)
        }

        if (!isOpening() && !isOpen()) {
          err = new errors.WriteError('Database is not open')
          return dispatchError(err, callback)
        }

        options  = getOptions(levelup, options)
        keyEnc   = options.keyEncoding
        valueEnc = options.valueEncoding

        arr = arr_.map(function (e) {
          if (e.type === undefined || e.key === undefined) {
            return {}
          }

          // inherit encoding
          var kEnc = e.keyEncoding || keyEnc
            , vEnc = e.valueEncoding || e.encoding || valueEnc
            , o

          // If we're not dealing with plain utf8 strings or plain
          // Buffers then we have to do some work on the array to
          // encode the keys and/or values. This includes JSON types.

          if (kEnc != 'utf8' && kEnc != 'binary'
              || vEnc != 'utf8' && vEnc != 'binary') {
            o = {
                type: e.type
              , key: util.encodeKey(e.key, options, e)
            }

            if (e.value !== undefined)
              o.value = util.encodeValue(e.value, options, e)

            return o
          } else {
            return e
          }
        })

        this.db.batch(arr, options, function (err) {
          if (err) {
            err = new errors.WriteError(err)
            return dispatchError(err, callback)
          } else {
            levelup.emit('batch', arr_)
            if (callback)
              callback()
          }
        })
      }

      // DEPRECATED: prefer accessing LevelDOWN for this: db.db.approximateSize()
      LevelUP.prototype.approximateSize = function (start_, end_, callback) {
        var err
          , start
          , end

        if (start_ === null || start_ === undefined
              || end_ === null || end_ === undefined
              || typeof callback != 'function') {
          err = new errors.ReadError('approximateSize() requires start, end and callback arguments')
          return dispatchError(err, callback)
        }

        start = util.encodeKey(start_, options)
        end   = util.encodeKey(end_, options)

        if (!isOpening() && !isOpen()) {
          err = new errors.WriteError('Database is not open')
          return dispatchError(err, callback)
        }

        this.db.approximateSize(start, end, function (err, size) {
          if (err) {
            err = new errors.OpenError(err)
            return dispatchError(err, callback)
          } else if (callback)
            callback(null, size)
        })
      }

      LevelUP.prototype.readStream =
      LevelUP.prototype.createReadStream = function (options) {
        options = extend(this.options, options)
        return readStream.create(
            options
          , this
          , function (options) {
              return levelup.db.iterator(options)
            }
        )
      }

      LevelUP.prototype.keyStream =
      LevelUP.prototype.createKeyStream = function (options) {
        return this.readStream(extend(options, { keys: true, values: false }))
      }

      LevelUP.prototype.valueStream =
      LevelUP.prototype.createValueStream = function (options) {
        return this.readStream(extend(options, { keys: false, values: true }))
      }

      LevelUP.prototype.writeStream =
      LevelUP.prototype.createWriteStream = function (options) {
        return writeStream.create(extend(options), this)
      }

      LevelUP.prototype.toString = function () {
        return 'LevelUP'
      }

      levelup = new LevelUP(location, options)
      levelup.open(callback)
      return levelup
    }

  , utilStatic = function (name) {
      return function (location, callback) {
        getLevelDOWN()[name](location, callback || function () {})
      }
    }

module.exports         = createLevelUP
module.exports.copy    = util.copy
// DEPRECATED: prefer accessing LevelDOWN for this: require('leveldown').destroy()
module.exports.destroy = utilStatic('destroy')
// DEPRECATED: prefer accessing LevelDOWN for this: require('leveldown').repair()
module.exports.repair  = utilStatic('repair')
